source/module_rdmft/cal_V_rdmft.cpp

```diff
@@ -0,0 +1,340 @@
+
+#include "rdmft.h"
+#include "module_rdmft/rdmft_tools.h"
+#include "module_psi/psi.h"
+#include "module_elecstate/module_dm/cal_dm_psi.h"
+
+#ifdef __EXX
+#include "module_ri/RI_2D_Comm.h"
+#include "module_hamilt_lcao/hamilt_lcaodft/operator_lcao/op_exx_lcao.h"
+#endif
+#include "module_hamilt_lcao/hamilt_lcaodft/operator_lcao/ekinetic_new.h"
+#include "module_hamilt_lcao/hamilt_lcaodft/operator_lcao/nonlocal_new.h"
+#include "module_hamilt_lcao/hamilt_lcaodft/operator_lcao/veff_lcao.h"
+
+namespace rdmft
+{
+
+
+
+template <typename TK, typename TR>
+void RDMFT<TK, TR>::get_DM_XC(std::vector< std::vector<TK> >& DM_XC)
+{
+    // get wk_funEta_wfc = wk*g(eta)*conj(wfc)
+    psi::Psi<TK> wk_funEta_wfc(wfc);
+    conj_psi(wk_funEta_wfc);
+    occNum_MulPsi(ParaV, wk_fun_occNum, wk_funEta_wfc, 0);
+
+    // get the special DM_XC used in constructing V_exx_XC
+    for(int ik=0; ik<wfc.get_nk(); ++ik)
+    {
+        // after this, be careful with wfc.get_pointer(), we can use &wfc(ik,inbn,inbs) instead
+        wfc.fix_k(ik);
+        wk_funEta_wfc.fix_k(ik);
+        TK* DM_Kpointer = DM_XC[ik].data();
+#ifdef __MPI
+        elecstate::psiMulPsiMpi(wk_funEta_wfc, wfc, DM_Kpointer, ParaV->desc_wfc, ParaV->desc);
+#else
+        elecstate::psiMulPsi(wk_funEta_wfc, wfc, DM_Kpointer);
+#endif            
+    }
+}
+
+
+template <typename TK, typename TR>
+void RDMFT<TK, TR>::cal_V_TV()
+{
+    HR_TV->set_zero();
+    
+    V_ekinetic_potential = new hamilt::EkineticNew<hamilt::OperatorLCAO<TK, TR>>(
+        hsk_TV,
+        kv->kvec_d,
+        HR_TV,
+        &GlobalC::ucell,
+        orb->cutoffs(),
+        &GlobalC::GridD,
+        two_center_bundle->kinetic_orb.get()
+    );
+
+    V_nonlocal = new hamilt::NonlocalNew<hamilt::OperatorLCAO<TK, TR>>(
+        hsk_TV,
+        kv->kvec_d,
+        HR_TV,
+        &GlobalC::ucell,
+        orb->cutoffs(),
+        &GlobalC::GridD,
+        two_center_bundle->overlap_orb_beta.get()
+    );
+
+    if( PARAM.inp.gamma_only )
+    {
+        V_local = new rdmft::Veff_rdmft<TK,TR>(
+            GG,
+            hsk_TV,
+            kv->kvec_d,
+            this->pelec->pot,
+            HR_TV,
+            &GlobalC::ucell,
+            orb->cutoffs(),
+            &GlobalC::GridD,
+            nspin,
+            charge,
+            rho_basis,
+            vloc,
+            sf,
+            "local"
+        );
+    }
+    else
+    {
+        V_local = new rdmft::Veff_rdmft<TK,TR>(
+            GK,
+            hsk_TV,
+            kv->kvec_d,
+            this->pelec->pot,
+            HR_TV,
+            &GlobalC::ucell,
+            orb->cutoffs(),
+            &GlobalC::GridD,
+            nspin,
+            charge,
+            rho_basis,
+            vloc,
+            sf,
+            "local"
+        );
+    }
+
+    // update HR_TV in ion-step, now HR_TV has the HR of V_ekinetic + V_nonlcao + V_local
+    V_ekinetic_potential->contributeHR();
```

Comment:
is the function `contributeHR` called here related with the one which looks quite similar in ABACUS LCAO implementation?